fix: allow spec.authorities field to not be specified

wojciechka opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Pleases use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

Closes #361 

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

Add `omitempty` to `authorities` to allow the `spec.authorities` field to not be specified.

This field being present and not nil is already conditionally validated (with the validation happening by default), so this should be safe to add.

To test the change:

1. Ensure `policy-controller` configuration includes `fail-on-empty-authorities: "false"`
2. Apply the following YAML:
```
apiVersion: policy.sigstore.dev/v1beta1
kind: ClusterImagePolicy
metadata:
  name: empty-authorities
spec:
  images:
  - glob: 'index.docker.io/*'
EOF
```

This should work, fails on latest main, but works with the code change from this PR.

#### Release Note

Allow `spec.authorities` field to not be specified

#### Documentation

N/A

hectorj2f commented:
You shared an error that is coming from the CRD definition where the type needs to be an array.

hectorj2f commented:
@wojciechka you need to signoff your commits

vaikas commented on pkg/apis/policy/v1alpha1/clusterimagepolicy_types.go:
```diff
@@ -55,7 +55,7 @@ type ClusterImagePolicySpec struct {
 	Images []ImagePattern `json:"images"`
 	// Authorities defines the rules for discovering and validating signatures.
-	Authorities []Authority `json:"authorities"`
+	Authorities []Authority `json:"authorities,omitempty"`
```
nit: Could we add the
```
// +optional
```

before L58

wojciechka replied:
Done in both files

vaikas commented on pkg/apis/policy/v1beta1/clusterimagepolicy_types.go:
```diff
@@ -55,7 +55,7 @@ type ClusterImagePolicySpec struct {
 	Images []ImagePattern `json:"images"`
 	// Authorities defines the rules for discovering and validating signatures.
-	Authorities []Authority `json:"authorities"`
+	Authorities []Authority `json:"authorities,omitempty"`
```
Ditto.

vaikas reviewed (commented):
Thanks for doing this! Could we also add two example test files here (one for v1alpha1 and one for v1beta1)?
https://github.com/sigstore/policy-controller/tree/main/test/testdata/policy-controller/valid

wojciechka commented:
> Thanks for doing this! Could we also add two example test files here (one for v1alpha1 and one for v1beta1)?
> https://github.com/sigstore/policy-controller/tree/main/test/testdata/policy-controller/valid

This is controlled by the `fail-on-empty-authorities` configuration option and not setting `spec.authorities` is not valid by default - you need to set `fail-on-empty-authorities` to `false` and only then it becomes possible.

There's even [test/testdata/policy-controller/invalid/v1beta1-keylessref-with-empty-authorities.yaml](https://github.com/sigstore/policy-controller/blob/main/test/testdata/policy-controller/invalid/v1beta1-keylessref-with-empty-authorities.yaml) that tests this is not valid `ClusterImagePolicy`.

hectorj2f commented:
@wojciechka I think we want to test the scenario we had when setting `fail-on-empty-authorities` to false where we weren't able to update a CIP. With these changes, we can now update an initial CIP without authorities.

wojciechka commented:
> I will suggest adding here the following test https://github.com/sigstore/policy-controller/blob/main/test/e2e_test_policy_crd.sh#L114

Thank you. Added the logic at the end of shell script and added test for both existing and non-existing `spec.authorities` in the YAML file.

hectorj2f reviewed (approved):
lgtm
